### PR TITLE
[REVIEW] Fix tail method issue for string columns

### LINKED ISSUE #2495: [BUG] tail method sometimes fail
After running a query, I am getting `ans` frame. `head` method works fine on it, but `tail` method fails. This happens rarely and strongly depends on the data. Using `0.8.0+0.g8fa7bd3.dirty`.
```
>>> ans = x.groupby(['id1'],as_index=False).agg({'v1':'sum'}).reset_index(drop=True)
>>> print(ans.head(3), flush=True)
     id1        v1
0  id001  15006850
1  id002  14994166
>>> print(ans.tail(3), flush=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/dataframe.py", line 491, in __str__
    return self.to_string(nrows=nrows, ncols=ncols)
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/dataframe.py", line 480, in to_string
    cols[h] = self[h].values_to_string(nrows=nrows)
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/series.py", line 354, in values_to_string
    out = [str(v) for v in values]
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/series.py", line 354, in <listcomp>
    out = [str(v) for v in values]
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/series.py", line 302, in __getitem__
    return self._column.element_indexing(arg)
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/column.py", line 412, in element_indexing
    val = self.data[index]  # this can raise IndexError
  File "/home/jan/anaconda3/envs/cudf/lib/python3.6/site-packages/cudf/dataframe/buffer.py", line 149, in __getitem__
    return item.view(self.dtype)
AttributeError: 'NoneType' object has no attribute 'view'
>>> ans.dtypes
id1    object
v1      int64
dtype: object
```
I can provide reproducible example but it will not be minimal... the one provided in https://github.com/rapidsai/cudf/issues/2494#issue-478107791 might work after changing K=2.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
The nvstrings was failing to extract the tail if the requested tail is bigger than length of the series which caused error somewhere else.
Example:
```
gdf = DataFrame()
gdf["id"] = Series(["a", "b"], dtype=np.object)
gdf["v"] = Series([1, 2])
print(gif.tail(-3))
```

```
length = 2

requested = 3
```
The code change is to fix `nvstrings` to provide proper result in this scenario. Test cases have been added for both `cudf` and `nvstrings`.

close #2495

### CODE REVIEW

File: python/nvstrings/nvstrings.py [hunk 1]
```diff
@@ -423,6 +423,8 @@ def __getitem__(self, key):
             end = self.size() if key.stop is None else key.stop
             step = 1 if key.step is None or key.step == 0 else key.step
             # negative slicing check
+            start = -1 * self.size() if start < -1 * self.size() else start
+            end = -1 * self.size() if end < -1 * self.size() else end
```
Comment (kkraus14) [inline python/nvstrings/nvstrings.py]: Should these just be modulo operations instead? It would greatly simplify this logic.

Comment (rgsl888prabhu) [reply python/nvstrings/nvstrings.py]: The following scenario would yield wrong result

```
start = -1
size = -2

size % start = 0
```
But expected is `-1`

Comment (kkraus14) [reply python/nvstrings/nvstrings.py]: @shwina how do we handle this normally for columns?

File: python/nvstrings/tests/test_substr.py [hunk 2]
```diff
@@ -24,6 +24,13 @@ def test_slice_from():
     assert_eq(got, expected)
 
 
+def test_slice_negative_start():
+    strs = nvstrings.to_device(["a", "b"])
+    got = strs[slice(-3, None, None)]
+    expected = ["a", "b"]
+    assert_eq(got, expected)
```
Comment (kkraus14) [inline python/nvstrings/tests/test_substr.py]: This test should go into a different file, this is for testing elementwise slicing of substrings.

File: python/nvstrings/nvstrings.py (hunk 1 above)
Comment (kkraus14) [reply python/nvstrings/nvstrings.py]: @rgsl888prabhu I think this should also handle positive numbers where they're larger than self.size(), yes?

Comment (rgsl888prabhu) [reply python/nvstrings/nvstrings.py]: It has been taken care in the cpp side of nvstrings.

File: python/nvstrings/tests/test_substr.py (hunk 2 above)
Comment (davidwendt) [reply python/nvstrings/tests/test_substr.py]: I would recommend moving this test to `test_array.py` .  It is closer in function to those methods (gather, sublist).

File: python/nvstrings/nvstrings.py (hunk 1 above)
Comment (shwina) [reply python/nvstrings/nvstrings.py]: We use the `indices` method of the slice. Chatted briefly with @rgsl888prabhu about it

**Comment by rgsl888prabhu** [comment]
@kkraus14 @davidwendt

**Comment by kkraus14** [comment]
rerun tests

File: python/nvstrings/nvstrings.py [hunk 3]
```diff
@@ -419,14 +419,7 @@ def __getitem__(self, key):
...
         if isinstance(key, slice):
-            start = 0 if key.start is None else key.start
-            end = self.size() if key.stop is None else key.stop
-            step = 1 if key.step is None or key.step == 0 else key.step
-            # negative slicing check
-            start = -1 * self.size() if start < -1 * self.size() else start
-            end = -1 * self.size() if end < -1 * self.size() else end
-            end = self.size() + end if end < 0 else end
-            start = self.size() + start if start < 0 else start
+            start, end, step = key.indices(self.size())
```
Comment (kkraus14) [inline python/nvstrings/nvstrings.py]: Didn't know this existed, sweet!

**Pull request merged**